contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,678 @@
+// SPDX-FileCopyrightText: 2023 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+/* See contracts/COMPILERS.md */
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts-v4.4/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts-v4.4/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts-v4.4/token/ERC20/extensions/draft-IERC20Permit.sol";
+import "@openzeppelin/contracts-v4.4/token/ERC20/utils/SafeERC20.sol";
+
+import {AccessControlEnumerable} from "@openzeppelin/contracts-v4.4/access/AccessControlEnumerable.sol";
+
+import "./lib/UnstructuredStorage.sol";
+
+/**
+ * @title Interface defining a Lido liquid staking pool
+ * @dev see also [Lido liquid staking pool core contract](https://docs.lido.fi/contracts/lido)
+ */
+interface IStETH {
+    /**
+     * @notice Get stETH token amount by the provided shares amount
+     * @param _sharesAmount shares amount
+     * @dev dual to `getSharesByPooledEth`.
+     */
+    function getPooledEthByShares(uint256 _sharesAmount) external view returns (uint256);
+
+    /**
+     * @notice Get shares amount by the stETH token amount
+     * @param _pooledEthAmount stETH token amount
+     * @dev dual to `getPooledEthByShares`.
+     */
+    function getSharesByPooledEth(uint256 _pooledEthAmount) external view returns (uint256);
+}
+
+interface IWstETH {
+    /**
+     * @notice Exchanges wstETH to stETH
+     * @param _wstETHAmount amount of wstETH to unwrap in exchange for stETH
+     * @dev Requirements:
+     *  - `_wstETHAmount` must be non-zero
+     *  - msg.sender must have at least `_wstETHAmount` wstETH.
+     * @return Amount of stETH user receives after unwrap
+     */
+    function unwrap(uint256 _wstETHAmount) external returns (uint256);
+
+    /**
+     * @notice Get amount of stETH for a given amount of wstETH
+     * @param _wstETHAmount amount of wstETH
+     * @return Amount of stETH for a given wstETH amount
+     */
+    function getStETHByWstETH(uint256 _wstETHAmount) external view returns (uint256);
+}
+
+/**
+ * @title A dedicated contract for handling stETH withdrawal request queue
+ * @author folkyatina
+ */
+contract WithdrawalQueue is AccessControlEnumerable {
```

Comment:
Maybe implement IERC721 (one of these fine days)

Reply:
Done 🎆